modules/context/repo.go

```diff
@@ -619,8 +619,8 @@ func RepoRefByType(refType RepoRefType) macaron.Handler {
 				// redirect from old URL scheme to new URL scheme
 				ctx.Redirect(path.Join(
 					setting.AppSubURL,
-					strings.TrimSuffix(ctx.Req.URL.String(), ctx.Params("*")),
-					ctx.Repo.BranchNameSubURL(),
+					path.Dir(ctx.Req.URL.String()),
+					strings.NewReplacer("%", "%25", "#", "%23", " ", "%20", "?", "%3F").Replace(ctx.Repo.BranchNameSubURL()),
```

Comment:
Please do not use custom escaping here, you can use go std lib escaping: `(&url.URL{Path: ctx.Repo.BranchNameSubURL()}).String()`

Reply:
Yeah that makes more sense. I'm not sure how and where `EscapePound` is used, otherwise it would maybe be good to   change that as well: [helper.go#L114](https://github.com/go-gitea/gitea/blob/535445c32ee730988033728b3b91c4d6f456e08c/modules/templates/helper.go#L114)